orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java

```diff
@@ -0,0 +1,175 @@
+   *     matches at most one artifact.
+   * @return An instance of {@link ArtifactResolver}
+   */
+  public static ArtifactResolver getInstance(
+      @Nullable Iterable<Artifact> currentArtifacts,
+      @Nullable Iterable<Artifact> priorArtifacts,
+      boolean requireUniqueMatches) {
+    return new ArtifactResolver(currentArtifacts, priorArtifacts, requireUniqueMatches);
+  }
+
+  private Optional<Artifact> resolveSingleArtifact(ExpectedArtifact expectedArtifact) {
+    Optional<Artifact> resolved = matchSingleArtifact(expectedArtifact, currentArtifacts);
+
+    if (!resolved.isPresent() && expectedArtifact.isUsePriorArtifact()) {
+      resolved = matchSingleArtifact(expectedArtifact, priorArtifacts);
+    }
+
+    if (!resolved.isPresent() && expectedArtifact.isUseDefaultArtifact()) {
+      resolved = Optional.ofNullable(expectedArtifact.getDefaultArtifact());
+    }
+
+    return resolved;
+  }
+
+  private Optional<Artifact> matchSingleArtifact(
+      ExpectedArtifact expectedArtifact, ImmutableList<Artifact> possibleArtifacts) {
+    ImmutableList<Artifact> matches =
+        possibleArtifacts.stream().filter(expectedArtifact::matches).collect(toImmutableList());
+
+    if (matches.isEmpty()) {
+      return Optional.empty();
+    }
+
+    if (matches.size() > 1 && requireUniqueMatches) {
+      throw new InvalidRequestException(
+          "Expected artifact " + expectedArtifact + " matches multiple artifacts " + matches);
+    }
+
+    return Optional.of(matches.get(0));
+  }
+
+  /**
+   * Resolves the input expected artifacts, returning the result of the resolution as a {@link
+   * ResolveResult}.
+   *
+   * <p>Resolving an expected artifact means finding an artifact that matches that expected
+   * artifact. To find a matching artifact, the following are considered in order:
+   *
+   * <ul>
+   *   <li>The {@link ArtifactResolver}'s current artifacts
+   *   <li>If the expected artifact has {@link ExpectedArtifact#isUsePriorArtifact()} true, the
+   *       {@link ArtifactResolver}'s prior artifacts
+   *   <li>If the expected artifact has {@link ExpectedArtifact#isUseDefaultArtifact()} true, the
+   *       {@link ExpectedArtifact}'s default artifact
+   * </ul>
+   *
+   * In order to determine whether an {@link ExpectedArtifact} matches an {@link Artifact}, the
+   * expected artifact's {@link ExpectedArtifact#matches(Artifact)} method is used.
+   *
+   * <p>If an expected artifact does not match any artifacts, an {@link InvalidRequestException} is
+   * thrown.
+   *
+   * <p>If {@link #requireUniqueMatches} is true, and an expected artifact matches more than one
+   * artifact in any of the above steps, an {@link InvalidRequestException} is thrown.
+   *
+   * @param expectedArtifacts The expected artifacts to resolve
+   * @return The result of the artifact resolution
+   */
+  public ResolveResult resolveExpectedArtifacts(
```

Comment:
very nitpicky, and let me know if this isn't how Spinnaker does it, but the _most common_ way to order Java methods is to put your public methods first, and your private methods close to the public method that uses them, like this:
* public void foo()
* private void calledByFoo()
* private void calledByCalledByFoo()
* public void bar()
* private void usedByBar()
* private void usedBySeveralPublicMethods()

This is completely backwards of idiomatic C++ style, I think.

Reply:
This :1st_place_medal: is going straight into my Java bible

Reply:
(Which is a google doc of stuff you and Eric have commented on PRs)

Reply:
Yeah, I don't think that Spinnaker has a particular well-defined order so definitely will follow this order instead of the random order I've been using.

Does that order include constructors?  So should I move above `public static ArtifactResolver getInstance` above the private constructor?  (I presume yes from the order above, but just checking there's no exception for constructors.)

Reply:
Hm, that's a good question. I would say I generally do see an exception for constructors and would expect a static factory method below them, like:

constants
fields
constructors
static factory methods (like `getInstance`)
API methods in the order described above

That's how I most often see them ordered, both inside and outside of Google. go/java-practices/class-structure has some more info. They actually do list "visible static methods" right below the constants (i.e. all the static stuff listed together), which doesn't feel particularly normal to me, but could definitely make sense! I don't know. I'd still be inclined to put `#getInstance` after the constructor still, but I'm not sure I'd feel super confident about it anymore 😁

Reply:
Makes sense!  Pushed a commit to update this.